adcc/LazyMp.py

```diff
@@ -60,39 +63,103 @@ def __getattr__(self, attr):
     def td2(self, space):
         """Return the T^D_2 term"""
-        return self.__cpp.td2(space)
+        if space != b.oovv:
+            raise NotImplementedError("T^D_2 term not implemented "
+                                      f"for space {space}.")
+        t2erit = self.t2eri(b.oovv, b.ov).transpose((1, 0, 2, 3))
+        denom = direct_sum(
+            'ia,jb->ijab', self.df(b.ov), self.df(b.ov)
+        ).symmetrise(0, 1)
+        return evaluate(
+            (+ 4.0 * t2erit.antisymmetrise(2, 3).antisymmetrise(0, 1)
+             - 0.5 * self.t2eri(b.oovv, b.vv)
+             - 0.5 * self.t2eri(b.oovv, b.oo)) / denom
+        )
 
+    @cached_member_function
     def t2eri(self, space, contraction):
         """
         Return the T2 tensor with ERI tensor contraction intermediates.
         These are called pi1 to pi7 in libadc.
         """
-        return self.__cpp.t2eri(space, contraction)
+        hf = self.reference_state
+        expressions = {
+            # space + contraction
+            b.ooov + b.vv: ('ijbc,kabc->ijka', b.ovvv),
+            b.ooov + b.ov: ('ilab,lkjb->ijka', b.ooov),
+            b.oovv + b.oo: ('klab,ijkl->ijab', b.oooo),
+            b.oovv + b.ov: ('jkac,kbic->ijab', b.ovov),
+            b.oovv + b.vv: ('ijcd,abcd->ijab', b.vvvv),
+            b.ovvv + b.oo: ('jkbc,jkia->iabc', b.ooov),
+            b.ovvv + b.ov: ('ijbd,jcad->iabc', b.ovvv),
+        }
+        contraction, eri_block = expressions[space + contraction]
+        return einsum(contraction, self.t2oo, hf.eri(eri_block))
 
-    @property
+    @cached_property
+    @timed_member_call(timer="timer")
     def mp2_diffdm(self):
         """
         Return the MP2 differensce density in the MO basis.
         """
-        ret = OneParticleOperator.from_cpp(self.__cpp.mp2_diffdm)
+        hf = self.reference_state
+        ret = OneParticleOperator(self.mospaces, is_symmetric=True)
+        ret[b.oo] = -0.5 * einsum("ikab,jkab->ij", self.t2oo, self.t2oo)
+        ret[b.ov] = -0.5 * (
+            + einsum("ijbc,jabc->ia", self.t2oo, hf.ovvv)
+            + einsum("jkib,jkab->ia", hf.ooov, self.t2oo)
+        ) / self.df(b.ov)
+        ret[b.vv] = 0.5 * einsum("ijac,ijbc->ab", self.t2oo, self.t2oo)
+
+        if self.has_core_occupied_space:
+            # additional terms to "revert" CVS for ground state density
+            ret[b.oo] += -0.5 * einsum("iLab,jLab->ij", self.t2oc, self.t2oc)
+            ret[b.ov] += -0.5 * (
+                + einsum("jMib,jMab->ia", hf.ocov, self.t2oc)
+                + einsum("iLbc,Labc->ia", self.t2oc, hf.cvvv)
+                + einsum("kLib,kLab->ia", hf.ocov, self.t2oc)
+                + einsum("iMLb,LMab->ia", hf.occv, self.t2cc)
+                - einsum("iLMb,LMab->ia", hf.occv, self.t2cc)
+            ) / self.df(b.ov)
+            ret[b.vv] += (+ 0.5 * einsum("IJac,IJbc->ab", self.t2cc, self.t2cc)
+                          + 1.0 * einsum("kJac,kJbc->ab", self.t2oc, self.t2oc))
```

Comment:
For some reason I don't like the + in front if it is not the first thing in the line .. but you decide, just a minor point.

Reply:
The + after ( ?
I can remove it but then it's not aligned with the `+ 1.0` in the next line...

Reply:
Resolved.